### PR TITLE
Staking UI

### PR DESCRIPTION
**Pull request by dansofter**
## Description

-Adds front end page to support staking on Calamari
-Adds front end support for multiple networks

---

Before we can merge this PR, please make sure that all the following items have been checked off. If any of the checklist items are not applicable, please leave them but write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work
- [x] Re-reviewed files changed in the Github PR explorer

## If PR to `staging`
- [ ] Updated relevant documentation in the code
- [ ] Test code changes manually, and describe your procedure in a comment on this PR
- [ ] Mark this PR with the correct milestone

## If PR to `main`
- [x] Update the version number in `package.json` and `src/config/common.json`
- [x] Update the minimum required signer version number in `src/config/common.json` if applicable
- [x] Test following the procedure in `docs/manual_test.md`
- [x] Update `CHANGELOG.md`
- [x] Make sure that all PR's to `manta-signer` and `sdk` on which this PR depends are merged

### CODE REVIEW

File: src/pages/StakePage/StakeContext/index.tsx [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+        .then((_unsub) => unsub = _unsub)
+        .catch(console.error);
+    };
+    setUserBalance();
+    let unsub;
+    return () => unsub && unsub();
+  }, [api, address, txStatus]);
+
+  useEffect(() => {
+    const getDelegations = async (api) => {
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:110]: @Kevingislason  move into queries file?

File: src/pages/StakePage/StakeContext/index.tsx [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+        ? delegatorState.value.delegations
+        : [];
+      return delegationsRaw.map(delegationRaw => {
+        const delegatedBalance = Balance.Native(new BN(delegationRaw.amount.toString()));
+        return new Delegation(delegationRaw.owner.toString(), delegatedBalance);
+      });
+    };
+
+    // todo: this is a disaster, refactor
+    const getUserRecentRewards = async (userDelegations) => {
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:122]: @Kevingislason  break apart as much as possible, comment, and describe in design doc

File: src/pages/StakePage/StakeContext/index.tsx [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+      for (const userDelegation of userDelegations) {
+        const collatorAddress = userDelegation.collator.address;
+        const topDelegationsRaw = await firstBlockApi.query.parachainStaking.topDelegations(
+          collatorAddress
+        );
+        const index = topDelegationsRaw.value.delegations
+          .map(delegation => delegation.owner.toString())
+          .indexOf(address);
+        if (index === -1) {
+          userDelegation.setRank(101); // i.e. > 100
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:199]: @Kevingislason  handle case where the user's delegation isn't active yet

File: src/pages/StakePage/StakeContext/index.tsx [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+
+    const getUserStakedBalance = (userDelegations) => {
+      return userDelegations.reduce(
+        (partialSum, delegation) => {
+          return partialSum.add(delegation.delegatedBalance);
+        },
+        Balance.Native(new BN(0)));
+    };
+
+    const setDelegatorState = () => {
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:214]: @Kevingislason think about function naming here

File: src/pages/StakePage/StakeContext/index.tsx [hunk 5]
```diff
@@ -0,0 +1,696 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:1]: refactor whole file

File: src/pages/StakePage/StakeContext/index.tsx [hunk 6]
```diff
@@ -0,0 +1,696 @@
...
+
+  const unstakeWouldBeBelowDelegationThreshold = () => {
+    const delegationThreshold = Balance.fromBaseUnits(
+      AssetType.Calamari(false), 5000
+    );
+    const balanceAfterUndelegation = selectedCollatorDelegation.delegatedBalance.sub(
+      unstakeTargetBalance
+    );
+    return balanceAfterUndelegation.lt(delegationThreshold);
+  };
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:423]: move validations into their own context

File: src/pages/StakePage/StakeContext/index.tsx [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+            );
+            const extrinsicHash = extrinsic.hash.toHex();
+            setTxStatus(TxStatus.finalized(extrinsicHash));
+          } catch(error) {
+            console.error(error);
+          }
+        }
+      }
+    }
+  };
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:491]: @Kevingislason  duplicated on send page

File: src/pages/StakePage/StakeContext/index.tsx [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+
+  const getUserCanWithdraw = async () => {
+    if (!userAvailableBalance || !api) {
+      return false;
+    }
+    const sufficientFunds = await getUserHasSufficientFundsToWithdraw();
+    return sufficientFunds;
+  };
+
+  const publishTx = async (tx) => {
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:585]: @Kevingislason  this is a general utility, move to utils

File: src/pages/StakePage/StakeContext/index.tsx [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+
+  const cancelUnstake = async () => {
+    const userCanCancelUnstake = await getUserCanCancelUnstake();
+    if (!userCanCancelUnstake) {
+      return;
+    }
+    const tx = buildCancelUnstakeTx();
+    await publishTx(tx);
+    setTxStatus(TxStatus.processing('Cancel transaction sent'));
+  };
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/index.tsx:629]: @Kevingislason  All these functions have same form, see about refactoring

File: src/pages/StakePage/StakeContext/stakeActions.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  SET_USER_BALANCE: 'SET_USER_BALANCE',
+  SET_USER_DELEGATIONS: 'SET_USER_DELEGATIONS',
+  SET_USER_RECENT_REWARDS: 'SET_USER_RECENT_REWARDS',
+  SET_USER_TOTAL_REWARDS: 'SET_USER_TOTAL_REWARDS',
+  SET_UNSTAKE_REQUESTS: 'SET_UNSTAKE_REQUESTS',
+  SET_TOTAL_REWARDS: 'SET_TOTAL_REWARDS',
+  SET_STAKE_TARGET_BALANCE: 'SET_STAKE_TARGET_BALANCE',
+  SET_UNSTAKE_TARGET_BALANCE: 'SET_UNSTAKE_TARGET_BALANCE',
+  SET_SELECTED_COLLATOR: 'SET_SELECTED_COLLATOR',
+  SET_SELECTED_UNSTAKE_REQUEST: 'SET_SELECTED_UNSTAKE_REQUEST',
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/stakeActions.ts:13]: @Kevingislason add SET_SELECTED_USER_DELEGATION

File: src/pages/StakePage/StakingTable.tsx [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    { field: 'Status', unSortIcon: true, width: 230, suppressMovable: true },
+    { field: 'Rank', unSortIcon: true, width: 230, suppressMovable: true },
+    {
+      field: '',
+      sortable: false,
+      width: 220,
+      suppressMovable: true,
+      cellRenderer: (params: any) => {
+        const delegation = params.data.data;
+        const unstakeRequest = unstakeRequests.find(
```
Comment (Kevingislason) [inline src/pages/StakePage/StakingTable.tsx:119]: @Kevingislason  This should be handled in stake context

File: src/pages/StakePage/index.tsx [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+import { SubstrateContextProvider } from 'contexts/substrateContext';
+import Navbar from 'components/Navbar';
+import { ExternalAccountContextProvider } from 'contexts/externalAccountContext';
+import DeveloperConsole from 'components/Developer/DeveloperConsole';
+import config from 'config';
+import { StakeContextProvider } from './StakeContext';
+import StakePageContent from './StakePageContent';
+
+const StakePage = () => {
+  console.log('config.DEV_CONSOLE', config.DEV_CONSOLE);
```
Comment (Kevingislason) [inline src/pages/StakePage/index.tsx:13]: get rid of

File: src/pages/StakePage/index.tsx [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  return (
+    <SubstrateContextProvider socket={'wss://crispy.baikal.testnet.calamari.systems'} >
```
Comment (Kevingislason) [inline src/pages/StakePage/index.tsx:15]: @Kevingislason This shouldn't be hard-coded

File: src/styles/ag-grid.scss [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+.ag-body-viewport {
+  background-color: var(--color-bg-secondary) !important;
+}
+
+.ag-row {
+  border: none;
+  height: 60px !important;
+  background-color: var(--color-bg-secondary) !important;
+  color: var(--color-text-secondary);
+  // border-bottom: 1px solid var(--color-border-secondary);
```
Comment (Kevingislason) [inline src/styles/ag-grid.scss:22]: delete

File: src/styles/ag-grid.scss [hunk 15]
```diff
@@ -0,0 +1,58 @@
+.ag-root-wrapper {
```
Comment (Kevingislason) [inline src/styles/ag-grid.scss:1]: @dansofter Can you also style the tooltip the ag-grid header tooltip text to be IBM Plex Mono?

File: src/styles/buttons.scss [hunk 16]
```diff
@@ -1,50 +1,65 @@
...
+  cursor: not-allowed;
+  background: rgba(0, 0, 0, 0.15) !important;
 }
 
 .btn-unreleased {
-    cursor: not-allowed;
-    background: rgba(0, 0, 0, 0.5) !important;
+  cursor: not-allowed;
+  background: rgba(0, 0, 0, 0.5) !important;
 }
```
Comment (Kevingislason) [inline src/styles/buttons.scss:41]: I think we can get rid of this

File: src/types/AssetType.ts [hunk 17]
```diff
@@ -38,6 +48,19 @@ export default class AssetType {
...
+      1,
+      'Calamari',
+      'KMA',
+      Svgs.Calamari,
+      12,
+      new BN('100000000000'),
+      isPrivate,
+      true
+    );
+  }
```
Comment (Kevingislason) [inline src/types/AssetType.ts:62]: @Kevingislason double check these values

File: src/types/Balance.ts [hunk 18]
```diff
@@ -42,18 +47,25 @@ export default class Balance {
     return valueBaseUnits;
   }
 
-  toString(shouldFormat) {
+  toString(shouldFormat, decimals = 3) {
```
Comment (Kevingislason) [inline src/types/Balance.ts:50]: remove default argument and specify everywhere

File: src/types/Collator.ts [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+  constructor(
+    address,
+    totalDelegation
+  ) {
+    this.address = address;
+    this.totalDelegation = totalDelegation;
+    this.name = this.mapAddressToName(address);
+    this.icon = null;
+    this.stats = null;
+    this.isActive = null; // todo: set to real value
```
Comment (Kevingislason) [inline src/types/Collator.ts:59]: remove comment

File: src/types/Collator.ts [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+  }
+}
+
+export class CollatorDelegationStats {
+  constructor(
+    balanceSelfBonded,
+    balanceEffectiveBonded,
+    delegationCount,
+    minStake
+  ) {
```
Comment (Kevingislason) [inline src/types/Collator.ts:91]: get rid of this type? or move total delegation into this type?

File: src/types/Delegation.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+    this.rank = null;
+  }
+
+  setRecentRewards(recentRewards) {
+    this.recentRewards = recentRewards;
+  }
+
+  setRank(rank) {
+    this.rank = rank;
+  }
```
Comment (Kevingislason) [inline src/types/Delegation.ts:19]: construct these classes all at once instead of using `set` methods

File: tailwind.config.js [hunk 22]
```diff
@@ -23,28 +23,31 @@ module.exports = {
     },
     extend: {
       backgroundImage: {
-        primary: 'var(--gradient-bg-primary)',
-        button: 'var(--gradient-button-primary)'
+        // primary: 'var(--gradient-bg-primary)'
+        // button: 'var(--gradient-button-primary)'
```
Comment (Kevingislason) [inline tailwind.config.js:27]: delete

**Review by Kevingislason — CHANGES_REQUESTED** [review]
@Kevin & @Danny: run eslint, fix errors, add prop types

@Kevin: Merge TS compilation PR

@Kevin: Comment functions

File: src/pages/StakePage/AccountSelect.tsx [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    return false;
+  };
+
+  useEffect(() => {
+    const timer = setTimeout(
+      () => addressCopied >= 0 && setAddressCopied(-1),
+      2000
+    );
+    return () => clearTimeout(timer);
+  }, [addressCopied]);
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/pages/StakePage/AccountSelect.tsx:38]: yeah it doesn't hurt

File: src/components/Navbar/ChainSelector.tsx [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+} from '@fortawesome/free-solid-svg-icons';
+import { useSubstrate } from 'contexts/substrateContext';
+import Svgs from 'resources/icons';
+import classNames from 'classnames';
+import { Link } from 'react-router-dom';
+
+
+const ChainDropdownItem = ({ node, activeNode }) => {
+  const { apiState } = useSubstrate();
+  const nodeIsDisconnected = apiState === 'ERROR' || apiState === 'DISCONNECTED';
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Navbar/ChainSelector.tsx:22]: On second thought I think this is okay

File: src/config/calamari.json [hunk 25]
```diff
@@ -1,84 +1,84 @@
...
+            "url": "wss://tempura.calamari.systems"
+        },
+        {
+            "name": "tender",
+            "url": "wss://tender.calamari.systems"
+        }
+    ],
+    "PAGE_TITLE": "Calamari App",
+    "PROVIDER_SOCKET": "wss://avocado.calamari.systems",
+    "SUBSCAN_URL": "https://calamari.subscan.io"
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/config/calamari.json]: Need to refactor handling of network configs

File: src/pages/BasePage.tsx [hunk 26]
```diff
@@ -0,0 +1 @@
+export const x = 1;
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/pages/BasePage.tsx:1]: too early for this

File: src/pages/StakePage/Modals/StakeModal.tsx [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+        if (targetBalance.valueAtomicUnits.gt(new BN(0))) {
+          setStakeTargetBalance(targetBalance);
+        } else {
+          setStakeTargetBalance(null);
+        }
+      } catch (error) {
+        console.error('error', error);
+      }
+    }
+  };
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/pages/StakePage/Modals/StakeModal.tsx]: seems non-trivial, will leave for another day

File: src/pages/StakePage/StakeContext/index.tsx (hunk 7 above)
Comment (Kevingislason) [reply src/pages/StakePage/StakeContext/index.tsx]: We can do this in a separate front end general cleanup PR

File: src/pages/StakePage/StakeContext/index.tsx (hunk 9 above)
Comment (Kevingislason) [reply src/pages/StakePage/StakeContext/index.tsx]: I think this is fine

File: src/pages/StakePage/StakingTable.tsx (hunk 11 above)
Comment (Kevingislason) [reply src/pages/StakePage/StakingTable.tsx]: not sure what I was thinking here, this seems fine

File: src/pages/StakePage/StakeContext/stakeActions.ts (hunk 10 above)
Comment (Kevingislason) [reply src/pages/StakePage/StakeContext/stakeActions.ts:13]: already covered by `selectedCollatorDelegation`

File: src/pages/StakePage/StakeContext/stakeReducer.ts [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+  return {
+    ...state,
+    userDelegations,
+    userStakedBalance,
+    userRecentRewards
+  };
+};
+
+const setUnstakeRequests = (state, { unstakeRequests }) => {
+  console.log('setUnstakeRequests reducer', setUnstakeRequests);
```
Comment (Kevingislason) [inline src/pages/StakePage/StakeContext/stakeReducer.ts:91]: get rid of

**Review by Kevingislason — CHANGES_REQUESTED** [review]
-add back signer connection status label
-add calamari production url
-min stake too high

**Review by Kevingislason — DISMISSED** [review]

**Review by AlexParshh — DISMISSED** [review]
looks good!

**Review by AlexParshh — APPROVED** [review]

**Pull request merged**